crates/ruff_linter/src/rules/ruff/rules/pytest_raises_ambiguous_pattern.rs

```diff
@@ -0,0 +1,161 @@
+use ruff_python_ast::{ExprCall, Keyword, StringLiteralValue};
+
+/// ## What it does
+/// Checks for non-raw literal string arguments passed to the `match` parameter
+/// of `pytest.raises()` where the string contains at least one unescaped
+/// regex metacharacter.
+///
+/// ## Why is this bad?
+/// The `match` argument is implicitly converted to a regex under the hood.
+/// It should be made explicit whether the string is meant to be a regex or a "plain" pattern
+/// by prefixing the string with the `r` suffix, escaping the metacharacter(s)
+/// in the string using backslashes, or wrapping the entire string in a call to
+/// `re.escape()`.
+///
+/// ## Example
+///
+/// ```python
+/// import pytest
+///
+///
+/// with pytest.raises(Exception, match="A full sentence."):
+///     do_thing_that_raises()
+/// ```
+///
+/// Use instead:
+///
+/// ```python
+/// import pytest
+///
+///
+/// with pytest.raises(Exception, match=r"A full sentence."):
+///     do_thing_that_raises()
+/// ```
+///
+/// Alternatively:
+///
+/// ```python
+/// import pytest
+/// import re
+///
+///
+/// with pytest.raises(Exception, match=re.escape("A full sentence.")):
+///     do_thing_that_raises()
+/// ```
+///
+/// or:
+///
+/// ```python
+/// import pytest
+/// import re
+///
+///
+/// with pytest.raises(Exception, "A full sentence\\."):
+///     do_thing_that_raises()
+/// ```
+///
+/// ## References
+/// - [Python documentation: `re.escape`](https://docs.python.org/3/library/re.html#re.escape)
+/// - [`pytest` documentation: `pytest.raises`](https://docs.pytest.org/en/latest/reference/reference.html#pytest-raises)
+#[derive(ViolationMetadata)]
+pub(crate) struct PytestRaisesAmbiguousPattern;
+
+impl Violation for PytestRaisesAmbiguousPattern {
+    #[derive_message_formats]
+    fn message(&self) -> String {
+        "Pattern passed to `match=` contains metacharacters but is neither escaped nor raw"
+            .to_string()
+    }
+
+    fn fix_title(&self) -> Option<String> {
+        Some("Use a raw string or `re.escape()` to make the intention explicit".to_string())
+    }
+}
+
+/// RUF043
+pub(crate) fn pytest_raises_ambiguous_pattern(checker: &mut Checker, call: &ExprCall) {
+    if !is_pytest_raises(&call.func, checker.semantic()) {
+        return;
+    }
+
+    let Some(Keyword { value, .. }) = call.arguments.find_keyword("match") else {
+        return;
+    };
+
+    let Some(string) = value.as_string_literal_expr() else {
+        return;
+    };
+
+    let any_part_is_raw = string.value.iter().any(|part| part.flags.prefix().is_raw());
+
+    if any_part_is_raw || !string_has_unescaped_metacharacters(&string.value) {
+        return;
+    }
+
+    let diagnostic = Diagnostic::new(PytestRaisesAmbiguousPattern, string.range);
+
+    checker.diagnostics.push(diagnostic);
+}
+
+fn string_has_unescaped_metacharacters(value: &StringLiteralValue) -> bool {
+    let mut escaped = false;
+    let mut last = '\x00';
```

Comment:
This can be simplified to 
```suggestion
    let last = value.chars().next_back().unwrap_or('\x00');
```

It's probably cheaper than assigning in every single iteration